Updated to a newer version of VTKm.

brugger1 opened the pull request:
### Description

Resolves #5293 

I upgraded VTKh and VTKm to more recent versions. VTKh was upgraded to the newest release v0.6.6 and VTKm was upgraded to version a3b852, which is the version that VTKh recommends.

This involved changes to build_visit to use the new versions of VTKh and VTKm. I was able to remove one of the patches applied to VTKm. The old version was over 2 years old and the API had changed a fair amount, so there were changes to the conversions between VTK and VTKm. Support for rectilinear grids had also been added in the intervening time, so the conversion to VTKm rectilinear grids was modified to use the native rectilinear grid instead of a uniform grid, which wasn't actually correct.

I also noticed the the tests for 2d curvilinear data were actually testing 3d curvilinear data, so I changed them to use 2d curvilinear data. I updated the baselines for those tests, since they were now testing something completely different. This was most likely a cut and paste error.

A few existing baselines had a single pixel difference, so they were updated.

Finally, the release notes were updated.

### Type of change

New feature.

### How Has This Been Tested?

I used build_visit to build the new versions for 3.2.0 on quartz. I then built VisIt using the new libraries and ran the test suite. I looked over the failures and determined the new images were correct and re-baselined those.

### Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have updated the release notes
- :x: I have made corresponding changes to the documentation
- :x: I have added debugging support to my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have added any new baselines to the repo
- [X] I have assigned reviewers (see [VisIt's PR procedures](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers) for more information)

brugger1 committed:
Updated to a newer version of VTKm.

brugger1 committed:
Merge branch 'develop' into task/brugger1/2020_12_04_vtkh_upgrade

markcmiller86 commented:
Fancy new checklist emojis...sweet!

biagas commented on src/avt/Filters/avtContourFilter.C:
```diff
@@ -537,6 +548,10 @@ avtContourFilter::ExecuteDataTree(avtDataRepresentation *in_dr)
+
+        //
+        // If we have a 2d structured grid, don't use VTKm.
+        //
```
Not part of your change, but I looked at the code for determining if the dataset is 2D, and it only looks at the z-dimension, what would happen if it z-dim > 1 but x or y isn't?  Should the logic be looking at the avtDataAttributes Spatial dimension instead?

brugger1 replied:
The conversion works on a vtkDataSet so I don't have access to avtDataAttributes here. I believe my changes will now detect 1d, 2d and 3d grids.

biagas replied:
The check for 2d rectilinear grid is:
``` 
         int dims[3];
            rgrid->GetDimensions(dims);

            if (dims[2] == 1)
            {
                useVTKm = false;
            }
```
I guess I am confused by what is truly meant by '2D', if dims were {20,1, 20}, isn't that 2D, just in a different plane than if dims were {20, 20, 1}?  VTKM handles the latter but not the former?

brugger1 replied:
Sorry, I thought I pushed the changes but didn't. Now it should really be fixed.

brugger1 replied:
I thought about this some more and don't think my changes are correct. I think my changes for the structured case are probably correct since I believe the indexing works out if nx=1 or ny=1. I don't think the same change for the rectilinear case is correct though. Was your concern only about structured grids?

biagas replied:
I was concerned about both cases, and whether or not passing data through VTKm would cause problems when dims[2] > 1, but either dims[0] or dims[1] was 1.  Does the testing still work with your most recent changes?

brugger1 replied:
The code will now skip VTKm if any of the dimensions is 1 for the rectlinear and structured case. I checked and VTKm doesn't even handle the 2D topology case if dims[2] == 1. It just gives a blank image.

biagas commented on src/avt/Pipeline/Data/avtDataRepresentation.C:
```diff
@@ -98,22 +98,54 @@ ConvertVTKToVTKm(vtkDataSet *data)
 
         int dims[3];
         rgrid->GetDimensions(dims);
+        int nDims = (dims[2] == 1 ? 2 : 3);
```
What happens if dims[0] or dims[1] is 1?

brugger1 replied:
I believe I've fixed this.

brugger1 replied:
I have no way to test this since VTKm doesn't work for 2D topology even in the case where dims[2] == 1. When it fails it results in an empty data set, resulting in a blank image, which isn't the worst failure mode. The logic for deciding if it is 2D is pretty conservative and only does it for the dims[2] == 1 case, leaving it in 3D for dims[0], dims[1] == 1. I have logic such that VisIt should never try to pass 2D data to a VTKm filter. Maybe one day VTKm will support 2D and I don't necessarily want to pull the code since I believe it is correct. Presumably anyone that pulls the tests for 2D would test that it works.

brugger1 committed:
Improve the code based on review comments.

brugger1 committed:
Merge branch 'task/brugger1/2020_12_04_vtkh_upgrade' of ssh://github.com/visit-dav/visit into task/brugger1/2020_12_04_vtkh_upgrade

brugger1 committed:
Allow zonal variables and 3D zoo elements.

brugger1 committed:
Update the build_visit_develop script for the Azure CI.

brugger1 committed:
Merge branch 'develop' into task/brugger1/2020_12_04_vtkh_upgrade

brugger1 committed:
Update the vmImage and CMake version in azure-pipelines.yml.

brugger1 committed:
Merge branch 'task/brugger1/2020_12_04_vtkh_upgrade' of ssh://github.com/visit-dav/visit into task/brugger1/2020_12_04_vtkh_upgrade

brugger1 commented:
I have added support for zonal variables in addition to nodal variables. I also allow 3D unstructured meshes with all the 3D zoo elements. I updated the baselines for the existing zonal variable tests (which used VTK) and added new ones for 3D zoo elements. All of these pass. I updated the release notes to reflect this as well.

brugger1 committed:
Merge branch 'develop' into task/brugger1/2020_12_04_vtkh_upgrade

brugger1 commented:
@biagas , @mclarsen this is ready for review again.

brugger1 committed:
Merge branch 'develop' into task/brugger1/2020_12_04_vtkh_upgrade
